### PR TITLE
OpenXLA pin update 07052023

### PR DESCRIPTION
**Pull request by lsy323**
XLA pin update:

XLA pin is updated to: `9b339c6fa10f6e964e21b58e40217661f7824bae`
Libtpu version it updated to: July 03 nightly

Patches:
- After gcc upgrade in #5297, patches for compilation error are removed(`cuda_graph.diff`, `triton_filesystem.diff`, `topk_rewriter`). Patches for back porting XLA features are removed. (`profiler_trace`, `xla_bzl`)
- Added a new patch (only 1 line, `constexpr_return`) for compilation error on gcc-11. The patch is needed as gcc-11 has more strict check.

Other changes:
- Update `GetStreamExecutorGpuClient` call due to API interface change, this only for GPU workflow. [Commit](https://github.com/openxla/xla/commit/b46de86747d37ddb16ec34cb9095bf0e196c223c) for API change on XLA side.
- Added TPU initialization logic in `PjRtComputationClient` ctor. XLA [commit](https://github.com/openxla/xla/commit/9409679b2945862e33c55dc7fee960caba50accd) in which TPU initialization logic is removed from `GetCApiClient`.
- Update `PartialTile` call and tile assignment logic to adapt to the API change in [PartialTile change](https://github.com/openxla/xla/commit/728e13fb733dba2e633bdac3af6d133aa419d545) and [OpSharding change](https://github.com/openxla/xla/commit/a7817135681c78734ad7bf4821b2436d970022f8#diff-a2231900a8321bb655a61ed7c8d3b8ffe98db70ed864c05d5c9f54a7b9e30a76) cc @alanwaketan and @jonb377 for this change.
- Added StableHLO version in `serializePortableArtifact` call, which is required now. 
- Added `--shm-size=16` when starting docker container for some GPU tests, which is required for ddp tests on large net, sync_batch_norm test.
- Minor refactor of `test_mp_sync_batch_norm.py`.

The following tests are disabled for now
- `test_operations.py:test_stack_pred` only fail on GPU, numerical result error.

### CODE REVIEW

**Comment by lsy323** [comment]
- CPU tests are passing
- GPU needs a proper fix for the `GetStreamExecutorGpuClient` API
- TPU CI failed due to an initialization error.

**Comment by lsy323** [comment]
Per offline discussion, attempted to update `GetStreamExecutorGpuClient` API by the following:
- Passing `PjRtComputationClient::GetNumProcesses()` to `num_nodes`. JAX is doing similar at https://github.com/google/jax/blob/b1985db27b9489d2cb051c69cbbb450d1f373876/jax/_src/xla_bridge.py#L287
- Leave nullptr for `kv_get` and `kv_put` as we don't exchange information across GPU hosts now.

Now CI failed at `PjRtComputationClient::GetNumProcesses()` call. Check with @will-cromar to see if there any insights before further investigating this.

cc  @vanbasten23  @JackCaoG

**Comment by will-cromar** [comment]
`GetNumProcesses` won't work until the underlying client is initialized, and now we need it to initialize the client. Instead, use `GPU_NUM_DEVICES` as the `num_nodes` since GPU only supports a single host today.

I expect that you will need `kv_put` and `kv_get` for exchanging information between processes.

File: .circleci/common.sh [hunk 1]
```diff
@@ -92,6 +92,19 @@ function install_deps_pytorch_xla() {
 
   sudo ln -s "$(command -v bazelisk)" /usr/bin/bazel
 
+  # Install gcc-8
+  sudo apt-get update
+  # Update ppa for GCC
+  sudo apt-get install -y software-properties-common
+  sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
+  sudo apt update -y
+  sudo apt install -y gcc-8
```
Comment (vanbasten23) [inline .circleci/common.sh:101]: Do you know why we need both gcc and g++?

**Comment by lsy323** [comment]
Verified ResNet 50 convergence with 10 epoch on v4-8, taking ~30min before and after pin update

**Comment by lsy323** [comment]
Opened an upstream PR with the PyTorch/XLA pin: https://github.com/pytorch/pytorch/pull/105896
All checks passed except one, but I think it's caused by upstream:
```
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 1842, in validate_model
    self.model_iter_fn(model, example_inputs)
  File "/var/lib/jenkins/workspace/benchmarks/dynamo/torchbench.py", line 437, in forward_pass
    return mod(*inputs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/torchvision/models/detection/generalized_rcnn.py", line 104, in forward
    proposals, proposal_losses = self.rpn(images, features, targets)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/torchvision/models/detection/rpn.py", line 372, in forward
    boxes, scores = self.filter_proposals(proposals, objectness, images.image_sizes, num_anchors_per_level)
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/torchvision/models/detection/rpn.py", line 288, in filter_proposals
    keep = box_ops.batched_nms(boxes, scores, lvl, self.nms_thresh)
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/torchvision/ops/boxes.py", line 75, in batched_nms
    return _batched_nms_coordinate_trick(boxes, scores, idxs, iou_threshold)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/jit/_trace.py", line 1221, in wrapper
    return fn(*args, **kwargs)
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/torchvision/ops/boxes.py", line 94, in _batched_nms_coordinate_trick
    keep = nms(boxes_for_nms, scores, iou_threshold)
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/torchvision/ops/boxes.py", line 41, in nms
    return torch.ops.torchvision.nms(boxes, scores, iou_threshold)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_ops.py", line 681, in __call__
    return self._op(*args, **kwargs or {})
RuntimeError: "nms_kernel" not implemented for 'BFloat16'
```

File: openxla_patches/constexpr_return.diff [hunk 2]
```diff
@@ -0,0 +1,12 @@
+diff --git a/xla/primitive_util.h b/xla/primitive_util.h
+index 696147844..dfea15a4d 100644
+--- a/xla/primitive_util.h
++++ b/xla/primitive_util.h
+@@ -748,6 +748,7 @@ inline bool FitsInIntegralType(int64_t x, PrimitiveType ty) {
+                  std::numeric_limits<NativeT>::max() >= x;
+         }
+         LOG(FATAL) << "Invalid primitive type " << PrimitiveType_Name(ty);
++	return false;
```
Comment (JackCaoG) [inline openxla_patches/constexpr_return.diff:9]: can we submit this patch to openxla directly?

File: torch_xla/csrc/runtime/env_vars.cc [hunk 3]
```diff
@@ -16,6 +16,7 @@ const char* const kEnvPjrtAsyncGpuClient = "PJRT_GPU_ASYNC_CLIENT";
 const char* const kEnvTpuLibraryPath = "TPU_LIBRARY_PATH";
 const char* const kEnvXpuLibraryPath = "XPU_LIBRARY_PATH";
 const char* const kEnvPjrtDistServiceAddr = "PJRT_DIST_SERVICE_ADDR";
+const char* const kEnvPjRtLocalProcessCount = "PJRT_LOCAL_PROCESS_COUNT";
```
Comment (JackCaoG) [inline torch_xla/csrc/runtime/env_vars.cc:19]: can we add this to https://github.com/pytorch/xla/blob/master/configuration.yaml

File: openxla_patches/constexpr_return.diff (hunk 2 above)
Comment (JackCaoG) [reply openxla_patches/constexpr_return.diff:9]: this is not blocking, just a question.

**Review by JackCaoG — APPROVED** [review]
LGTM in general, I will let @will-cromar to take another look at the GPU change since I am not familiar with that logic.

**Review by will-cromar — APPROVED** [review]
LGTM. Thanks!

File: torch_xla/csrc/runtime/env_vars.cc (hunk 3 above)
Comment (lsy323) [reply torch_xla/csrc/runtime/env_vars.cc:19]: This var shouldn't be expected to be set by users, so I guess we don't export to the configuration.yaml? It's set here https://github.com/pytorch/xla/blob/74d115ec778c2463ca422d800a0948bc62780f20/torch_xla/_internal/pjrt.py#L109

File: openxla_patches/constexpr_return.diff (hunk 2 above)
Comment (lsy323) [reply openxla_patches/constexpr_return.diff:9]: Yes, we should do that..Let me create a cl later

**Pull request merged**